File: Javascript/partitionList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} x
 * @return {ListNode}
 */
var partition = function (head, x) {
  const before_head = new ListNode(0, null);
  const after_head = new ListNode(0, null);
  let before = before_head;
  let after = after_head;
  while (head) {
    if (head.val < x) {
      before.next = head;
      before = before.next;
    } else {
      after.next = head;
      after = after.next;
    }
    head = head.next;
  }
  after.next = null;
  before.next = after_head.next;
  return before_head.next;
};

File: Python/leetcode/Partition_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
#         create two dummy nodes for two parts of the answer before and after
        before=before_head=ListNode(0,None)
        after=after_head=ListNode(0,None)
#         Iterate over the given list while adding the node<x to before list and adding node>=x to after list
        while head:
            if head.val<x:
                before.next=head
                before=before.next
            else:
                after.next=head
                after=after.next
            head=head.next
#             Set after.next to point at null since this is the end of list
        after.next=None
#     Set before.next to after_head.next to join the two parts
        before.next=after_head.next
#     Return before_head.next as answer
        return before_head.next